src/H5Dio.c

```diff
@@ -323,7 +338,11 @@ H5D__read(size_t count, H5D_dset_io_info_t *dset_info)
         }
 
         /* MDIO-specific second phase initialization */
-        for (i = 0; i < count; i++)
+        for (i = 0; i < count; i++) {
+            /* Check for skipped I/O */
+            if (dset_info[i].skip_io)
```

Comment:
Make sure we don't call the `mdio_init` callback for datasets where we skipped I/O, because the callbacks may try to use fields that aren't initialized for the datasets